### ISSUE TITLE
Minimum Viable Product approach vs Alpha - Beta - Open Source

### ISSUE THREAD

**Issue by lisafast** [canada-ca/Open_First_Whitepaper#17]
A Lean iterative Minimum Viable Product approach is not the same as the Alpha - Beta model described in the Principles. An Alpha - Beta usually assumes an entire service rolls out at the same time. Rather than rolling out small experiments, it rolls out the service as an Alpha and so on. 

Here's the [classic MVP image by Henrik Kniberg](http://blog.crisp.se/2016/01/25/henrikkniberg/making-sense-of-mvp), modified for government megaprojects: 
![minimal-viable-product-phoenix-kniberg](https://user-images.githubusercontent.com/28159252/32112891-90c09e82-bb0c-11e7-96f7-e90a67f6cce0.jpg)

  I was recently asked to respond to this question about MVP's, and thought I would post the response here for this white paper.   

**Q1. If you are the company coming in AFTER the MVP, what do you need to pick up and get past the incumbency barriers?  What are the conditions for success?**
 
Only government would ask this question – no start-up would dream of switching out the entire dev team after a first successful iteration! Would a company GROW the team? Yes, absolutely. Over time, as the iterations proceed and the product grows and succeeds, new team members come on and some team members move out – gradually. But given the expense and difficulty of finding good developers, no company would switch them out, especially if that team is really rocking the process. 

However, the idea of keeping a core team does provide one answer. Success, and particularly for Open Source projects, requires having an experienced internal Product Owner/Product Manager in place  – as per UK role description: [UK Government Digital Services Service manual](https://www.gov.uk/service-manual/the-team/what-each-role-does-in-service-team)
 
In the Lean Start-Up Model, every single iteration is an MVP, each one is just enough for a valid experiment building on previous iterations. So the answer to your question totally depends on the level of investment in the previous iteration. 
 
- In the eRegs case, where the first MVP was a small $20K contract, the code base is large but the infrastructure around that in terms of documentation and all of the things @mgifford outlines here  https://github.com/mgifford/open-source-contracting just isn’t there yet. And why should it be – we were asked to build a skateboard, we built a bicycle, and neither comes with a manual.  It will take another iteration or two to get to that point, and again, it will depend on the requirements for that iteration. If the client asks for a LOT of new features in an iteration (which is not how it should be done but …government ) and still only wants to spend $20K, then once again, those open source requirements won’t be met, and it will be tough to get the code into condition for someone else to pick up. 

This is especially true if there are small projects with no funding in between to maintain the code. Open source code relies on other platforms - they evolve, and your code has to do so too. 
 
This company tells the same story: https://blog.wearefuturegov.com/lets-open-source-by-default-fb06f509dd9e

> “We recognise that it would be a lot for someone to take on the code base as it stands, in such an early prototype state. The limited time to deliver an MVP, leads to the code being less robust than we’d like. However, by sharing the project progress and research we hope it will be useful to any council considering a similar technical tool for their ways into work programme.”

It would also make sense to take this in steps – first step is to have the incumbent work with another company as a contributor on the project. That process will reveal any gaps and fill them in. By the next iteration, the incumbent could be ready to clear out and get an entirely new team.
 
Summary: The conditions for success are based on the investment in the MVP. If it’s a quick ‘test the waters’ by building just the skateboard  for very little money and a tiny temporary internal team, then it simply can’t be successful to transition. If you are on the 3 or 4th iteration,  have all of the open-source systems set up, have a solid Product Manager in place who can carry and manage the knowledge and backlog, then you are ready.

**Comment by mgifford** [comment]
Thanks for this @lisafast - what it boils down to is that setting up any ecosystem takes time and attention. In general I'd love to see government see if they can do whatever possible to not start a new software project.  Is there any way that government can leverage what already exists to do what they are trying to accomplish? If it is at all possible to build on someone else's work there is a lot more opportunity to have accessible, secure, performant and beautiful technology. 

What I would like to see is:
0 - A study to see that existing code or project can't be leveraged.
1 - A simple MVP of a generic tool with a single vendor, released under a pervasive open source license
2 - Government engage with, modify & customize that MVP to meet their needs, hiring the vendor that did the initial development. Government would be contributing documentation, specific use case and additional testing
3 - Government engage with a second vendor to work in conjunction with the first vendor to provide enhancements that would help more general adoption. 
4 - Government promotes solution to other departments and agencies (ideally internationally). Vendors 1 & 2 get contracts building sites for new clients & the software is further generalized to meet a broader range of criterion.
5 - More vendors are brought into the cycle of developing and extending the original MVP and it becomes used by more than government.

People tend to forget that **all code is debt**. Secure, accessible, usable, sustainable, code requires a community to maintain it. 

What you don't want is to for government to hire a great local firm to develop a MVP (which they will probably have to do at a loss), have them open source it only to find that that department is trying to get it customized on the cheap by interns or freelancers rather than actually investing in building a solid ecosystem.

**Comment by lisafast** [comment]
@mgifford Thank you for this sequence.  Note that it is not an Alpha -Beta model, as outlined in the principles of the white paper. That first step of finding code to be leveraged is super-important - @steve-h suggests that this is in itself a development job - and requires a certain level of competency to do it, as well as a good understanding of the problem to be solved and the long term vision. 

The point is you are not looking for the easiest first step, but one that highlights the path, AND allows the team to test their assumptions (as in the above image with the skateboard), before building any further. 

BTW, the numbered sequence quite accurately reflects what happened/is happening with eRegs so far, except for "1. A simple MVP of a generic tool". It is often much more difficult to build a generic tool than to build a specific one that will enable you to test assumptions with real users. In an MVP, you rough in/stub the features that later will allow you to make it generic.  Really you are talking top-down vs bottom-up. If the problem space is well-understood, bottom up can work. But in many cases,  a LEAN MVP approach is needed specifically because the space is not well understood. 

@steve-h says for example, it would be easier to write a specific cloud-deployment script and generalize it, rather than starting from some generic point. In our case, we built an MVP pilot for a specific regulatory amendment from a specific source - the Canada Gazette. It launched on time, and allowed us to test many design assumptions with the external users and internal analysts. The goal of the next iteration is to get more generic.

**Comment by mgifford** [comment]
Thanks @lisafast Definitely agree with you & @steve-h that having a clear set of broad goals & and basic requirements is something that could be carved off into it's own project. Also agree with you about the ease of creating an initial specific target and then attempting to generalize it. This often happens where developers have the option to solve a problem and then refactor how they did it so that it can be managed with smaller more independent parts. 

The goal has to be about meeting broader needs though so that it is worth releasing the code. If the project never moves beyond the needs of one department then it probably shouldn't be sitting up on a public repo in the first place. So how's something like this:

Government should be:
0 - Engaging a study the problem and investigate if there is any existing code or projects that can be leveraged. The less code one needs to maintain, the better. 
1 - Building a quick MVP with a single vendor and license it under a pervasive open source license. Make sure it works and that there is value in re-purposing the code.
2 - Working with that vendor to produce a more generic & better documented version of the code and release it publicly. 
2 - Engaging with, modify & customize that MVP to meet their needs, hiring the vendor that did the initial development. Government would be contributing documentation, specific use case and additional testing
3 - Bringing in a second vendor to work in conjunction with the first vendor to provide enhancements that would help more general adoption.
4 - Promoting solution to other departments and agencies (ideally internationally). Vendors 1 & 2 get contracts building sites for new clients & the software is further generalized to meet a broader range of criterion.
5 - Enlisting more vendors into the cycle of developing and extending the original MVP and it becomes used by more than government.

**Comment by lisafast** [comment]
This is great! Thanks @mgifford

**Comment by petersilva** [comment]
People do tend to start too many projects, and there is, in general, many instances of code that does similar tasks, and there should be greater emphasis on re-use and collaboration. Even so, I find the blanket statement **all code is debt** to be a little strong. Code is an asset, and it requires maintenance, just like a car is an asset that requires maintenance. Without maintenance that asset can depreciate to less than zero value (garage space for a car that doesn´t run, and can´t even give away.) but cars are not just debt.  The asset provides value that makes the maintenance worthwhile, and using existing code is about minimizing our maintenance costs. Unlike cars (ok forget car sharing for a minute) we can share code, and maintain it together. Minimizing code is a good goal in general, and in practice, we should get as much functionality as possible out of as little code as possible. Code that enables us to do things that we cannot do without it is a positive asset, that must be maintained to retain it´s value, or minimize depreciation.

One way that code is different from cars is that if the more it is used, the more valuable it is (in that it provides value to users) and if maintenance activities make it more usable and adaptable, it becomes suitable for more people to use, a virtuous cycle where maintenance costs causes appreciation.

I´m not sure where this is going, but **all code is debt** while often good advice in practice, seems overly bleak.

**Comment by RussellMcOrmond** [comment]
@petersilva , I am generally uncomfortable with analogies between intangibles such as software and tangibles such as cars.  If comparing to anything in transportation, software is analogous to knowledge held by skilled drivers and not the car.

For the government I believe the most constructive analogy is to other policy.  Policy is expensive to develop and maintain, but must be maintained to continue to have value (where the 'debt' part comes in).  Policy has value, but is quite appropriately not seen by government as an "asset" and is not seen as part of the "procurement" process.  The value of policy increases as it is shared, and disclosure is often necessary for the policy to be effective: Laws must be able to be read and be able to be made understandable by citizens to be obeyed, otherwise the legal system can't work.

Maybe the word "debt" feels too strong, but I believe thinking of software code as comparable to tangible assets is far more problematic.

We have had interests who for decades sold society on the idea that thinking of software and other intangibles as having the limitations of tangibles is a useful construct.  I see [Gates open letter to Hobbyists in 1976](https://en.wikipedia.org/wiki/Open_Letter_to_Hobbyists) as the launch of that campaign, even though software wasn't recognised in many countries under copyright law until the 1980's.  Before that time software was thought of as knowledge and instructions for operating the computer, commonly shared between operators, and not as a separate "product" or "asset".

Charging royalties for access to instructions is understood as merely a business model, not an inherent feature of educational material.  People generally think of teaching as an honourable activity, not as "stealing", even if the educational material was expensive to develop.

I believe part of the policy discussions with governments discussing what it means to be "open" in the context of software is to finally move away from the flawed thinking of software as being "a car" rather than being "driving instructions".

**Comment by sboots** [comment]
@RussellMcOrmond this is a great point:

> I believe part of the policy discussions with governments discussing what it means to be "open" in the context of software is to finally move away from the flawed thinking of software as being "a car" rather than being "driving instructions".

Well-said on all points!

**Comment by petersilva** [comment]
if we think of it as driving instructions, then code is "knowlege" so **all code is debt** becomes **all knowledge is debt**, the less we know, the better off we are?  I see that as problematic as an absolute, but often of practical help.   We should apply Occam's Razor **the simplest explanation is usually the correct one** , but also be mindful of Einstein's formulation: **Everything should be made as simple as possible, but not simpler.**

**Comment by RussellMcOrmond** [comment]
@petersilva The "* is debt" conversation comes out of a specific concern knowledge workers have of management thinking of this codified knowledge as assets that once "procured" will continue to solve future problems.   I agree it can't be seen as an absolute, but only as a response to that specific management problem.

It has been years since I have authored software in 6502 assembly (I'm 13 years younger than Bill Gates, so don't have software experience from the 1970's, just starting in the early 1980's), but I have constantly upgraded my software authoring and related computer operation knowledge over the years.  Most (but not all) of the software I authored today was in Perl.  I don't expect these specific lines of code to still be used a decade from now, even if some of the concepts embedded in them are carried forward.  I would be very concerned if a manager thought, "We have created this feature, and now we will be able to do X forever without ever having to write software again".

That assembler I wrote in high-school, and that BBS I presented at the Canada-wide Science Fair, hasn't been of much value to anyone (beyond archivists and archaeologists :-) since the 1980's.  Maintenance wouldn't have been enough, as the whole need for that specific written-down knowledge no longer exists.
